README.md
dcrweb
======

This is the code for the [decred.org](https://www.decred.org) website.

[![Build Status](https://travis-ci.org/decred/dcrweb.png?branch=master)](https://travis-ci.org/decred/dcrweb)
[![ISC License](http://img.shields.io/badge/license-ISC-blue.svg)](http://copyfree.org)


#### Downloading web site code

Make sure you have [git](https://git-scm.com/) installed.

```sh
git clone https://github.com/decred/dcrweb
```

## Development

1. Install prerequisites:

   * [Hugo](https://gohugo.io/)
   * Docker

2. Clone repo
    ```sh
    git clone https://github.com/decred/dcrweb
    cd dcrweb
    ```

3. Start development web server:
    ```sh
    bin/watch.sh
    ```

You should now be able to access the site at `http://localhost:1313`

#### Editing content

`dcrweb` is built using the [Hugo](https://gohugo.io/) site generator framework.  The most frequently updated content sections live in the below locations:


| Section | File | Comments |
| --- | --- | --- | 
| Press releases | `src/content/news/*.md` | When adding a new release, please follow the file naming convention in the directory.
| Press coverage | `src/data/news/coverage.yml` ||
| Download links | `src/data/wallets/links.yml` ||
| Current release | `src/data/wallets/current_release.yml` | The current release as it appears in the footer|
| Contributors      |  `src/data/contributors/*.yml` | Avatar images: `src/assets/images/contributors` |
| Community channels    | `src/data/community/channels.yml` |Logo images: `src/assets/images/community` |
| Exchanges             | `src/data/exchanges.*.yml` | Logo images: `src/assets/images/exchanges`|

The other sections live under `src/layouts`.  These pages are implemented as [Hugo templates](https://gohugo.io/templates/) and are [localized](https://gohugo.io/content-management/multilingual/#translation-of-strings).  The message catalogs can be found in `src/i18n`, when making changes in the templates, you'll want to keep the strings in the catalogs, please follow the naming scheme in the existing templates.    

## Testing

Run the HTML validator to make sure all of the generated files are syntactically correct.  (The script depends on `yarn` and `docker` being installed.)


```sh
yarn && yarn test
```


## Localization

The below commands must be run when either the content changes or there are updates in the translations in Transifex.  You'll first need to install the [Transifex client](https://docs.transifex.com/client/installing-the-client).


#### Importing new translations and content updates

When translations are added/updated in [Transifex](https://www.transifex.com/decred/), pull the updates:

```sh
bin/transifex_pull.sh
```

When you run this for the first time, you'll be asked to log in with your Transifex username/password.

To push the changes to staging:

```sh
git commit -m'Translation update'
git push origin
```

#### Updating the message catalog

When the master content changes in the HTML files, you'll need to re-generate the message catalog and push it to Transifex so that translators can update the localized message catalogs:

```sh
bin/transifex_push.sh
```

## Deployment

A Docker configuration is included for building the deployable images of dcrweb.

### Prerequisites
  - docker

### Building the Docker image for deployment

```sh
bin/build.sh
```
This builds the docker image `decred/dcrweb`, which can then be run using:

```sh
docker run -d -p <local port>:80 decred/dcrweb:latest
```

### Push to Dockerhub

```sh
docker login
```

Enter your [Docker HUB](https://hub.docker.com/) credentials that has write access to the `decred/dcrweb` repository.

```sh
docker push decred/dcrweb
```

### Run in production

```sh
docker pull decred/dcrweb
docker run -d --rm -p <local port>:80 decred/dcrweb:latest
```

Runs the docker image exposing the specified local port.

## License

dcrweb is licensed under the liberal ISC License.

drafts/history.md
# History

### Decred Project Origin

The seeds of Decred were planted in the early years of Bitcoin. In April of 2013, on Bitcointalk, a developer named tacotime introduced the idea of a cryptocurrency that actually changes something, based upon the Memcoin2 (MC2) paper by Adam Mackenzie. Tacotime spent most of 2013 working with another entity, _ingsoc, in an effort to fund the project. By early 2014, these efforts brought the idea to Company Zero (c0), a group of privacy- and security-oriented cypherpunks who wrote btcsuite in Golang, the first alternative full node implementation of Bitcoin.

The three entities worked together to create a fairer cryptocurrency based upon MC2 that empowered the coin holders with direct sovereignty and provided continuous funding. In late 2015, Jake Yocom-Piatt of c0 published articles describing Bitcoin’s biggest challenges, and introducing Decred to the world. Three days later tacotime submitted their last ever post on the MC2 thread, “The descendant of this system is here:– Closing this thread.”

### Airdrop & Premine

When launched in early 2016, Decred used an innovative airdrop and premine to bootstrap the Proof-of-Stake component of the Decred network, establish a wide network of potential project contributors, and compensate the early developers for their work prior to launch. Given the necessity of distribution, this process was determined to be the most fair and transparent, and yielded significant decentralization from the outset.

## Decred Timeline

2015-12-15 Decred announced on bitcointalk. -
https://bitcointalk.org/index.php?topic=1290358.0

2016-01-25 Airdrop vetting finalized. -
https://forum.decred.org/threads/airdrop-rundown.313/
2016-02-08 Decred mainnet launched. -
https://explorer.dcrdata.org/block/
298e5cc3d985bfe7f81dc135f360abe089edd4396b86d2de66b0cef42b21d98
0
2016-02-21 First PoS votes are included in block 4096. -
https://explorer.dcrdata.org/block/
00000000000013722f8e5a8af9cf55492e9237e77d29da98695e65fd1303362
5
2016-07-25 First release of Paymetheus Windows GUI wallet. -

2016-12-26 First release of Decrediton GUI wallet for Linux/OSX. -
https://forum.decred.org/threads/dd-20-v0-7-0-12-26-16.4702/#post-23300


2017-07-09 First hard fork vote in the history of cryptocurrency to
change Decred's stake difficulty algorithm. - https://blog.decred.org/

2017-07-09 On-chain signaling vote to begin work on Lightning Network support.
2017-09-19 First cross-chain atomic swap between Decred and Litecoin. - https://blog.decred.org/2017/09/20/On-Chain-Atomic-Swaps/
2017-10-25 Politeia is announced. - https://blog.decred.org/2017/10/25/Politeia/
2018-06-05 DEX proposal is published. - https://blog.decred.org/2018/06/05/A-New-Kind-of-DEX/
2018-09-21 Initial SPV release for dcrwallet. - https://github.com/decred/decred-binaries/releases/tag/v1.3.0
2018-10-15 Politeia goes live on mainnet. - https://blog.decred.org/2018/10/15/Politeia-in-Production/
2019-08-28 Initial Privacy Implementation Released https://blog.decred.org/2019/08/28/Iterating-Privacy/

## Decred Community Principles

These principles guide the Decred DAO as it builds a community-directed cryptocurrency whose security, adaptability, and sustainability make it a superior long-term store of value.

Free and Open-Source Software
All software developed as part of Decred shall be free and open-source software.
Incremental Privacy and Security
Privacy and security technology shall be implemented on a continuing and incremental basis, both proactively and on-demand in response to attacks.
Free Speech and Consideration
Everyone has the right to communicate opinions and ideas without fear of censorship. Consideration shall be given to all constructive speech that is based in fact and reason.
Multi-Stakeholder Inclusivity
Inclusivity represents a multi-stakeholder system and an active effort shall be maintained to include a diverse set of views and users.
Fixed Finite Supply - Issuance is finite and the total issuance shall not exceed 20,999,999.99800912 DCR
Universal Fungibility - Universal fungibility is fundamental to Decred being a store of value and attacks against it shall be actively monitored and countermeasures pursued as necessary.

——————————————————————————————
### Find out more about

# Adaptability
Learn how Decred empowers its stakeholders with direct sovereignty to manage the future of their money.

# Security
Learn how the Hybrid PoW/PoS system layers security to make Decred far more secure than other cryptocurrencies.

# Sustainability
Learn about Decred’s long horizon, it’s self-funding mechanism, and how the DAO enables a new work paradigm.

# History
Learn about the origins of Decred and key points in its history.


api/index.php
<?php

$GLOBALS["clearcache"] = 0;
$GLOBALS["debug"] = 0;

define("STAKEPOOL_API_INITIAL_VERSION", 1);
define("STAKEPOOL_API_CURRENT_VERSION", 2);

if ($GLOBALS["clearcache"]) {
    apcu_clear_cache();
}

// header('Access-Control-Allow-Origin: *');

$spdata = array(
    "Bravo" => array(
        //"LaunchedEpoch" => strtotime("Sun May 22 17:54:00 CDT 2016"),
        "APIEnabled" => false,
        "APIVersionsSupported" => array(),
        "LastAttempt" => 0,
        "LastUpdated" => 0,
        "Network" => "mainnet",
        "URL" => "https://dcr.stakepool.net",
    ),
    "Delta" => array(
        //"LaunchedEpoch" => strtotime("Thu May 19 10:19:00 CDT 2016"),
        "APIEnabled" => false,
        "APIVersionsSupported" => array(),
        "LastAttempt" => 0,
        "LastUpdated" => 0,
        "Network" => "mainnet",
        "URL" => "https://dcr.stakeminer.com",
    ),
    "Echo" => array(
        //"LaunchedEpoch" => strtotime("Mon May 23 12:59:00 CDT 2016"),
        "APIEnabled" => false,
        "APIVersionsSupported" => array(),
        "LastAttempt" => 0,
        "LastUpdated" => 0,
        "Network" => "mainnet",
        "URL" => "https://pool.d3c.red",
    ),
    "Foxtrot" => array(
        //"LaunchedEpoch" => strtotime("Tue May 31 08:23:00 CDT 2016"),
        "APIEnabled" => false,
        "APIVersionsSupported" => array(),
        "LastAttempt" => 0,
        "LastUpdated" => 0,
        "Network" => "mainnet",
        "URL" => "https://dcrstakes.com",
    ),
    "Golf" => array(
        //"LaunchedEpoch" => strtotime("Wed May 25 04:09:00 CDT 2016"),
        "APIEnabled" => false,
        "APIVersionsSupported" => array(),
        "LastAttempt" => 0,
        "LastUpdated" => 0,
        "Network" => "mainnet",
        "URL" => "https://stakepool.dcrstats.com",
    ),
    "Hotel" => array(
        //"LaunchedEpoch" => strtotime("Sat May 28 14:31:00 CDT 2016"),
        "APIEnabled" => false,
        "APIVersionsSupported" => array(),
        "LastAttempt" => 0,
        "LastUpdated" => 0,
        "Network" => "mainnet",
        "URL" => "https://stake.decredbrasil.com",
    ),
    "India" => array(
        //"LaunchedEpoch" => strtotime("Sun May 22 13:58:00 CDT 2016"),
        "APIEnabled" => false,
        "APIVersionsSupported" => array(),
        "LastAttempt" => 0,
        "LastUpdated" => 0,
        "Network" => "mainnet",
        "URL" => "https://stakepool.eu",
    ),
    "Juliett" => array(
        //"LaunchedEpoch" => strtotime("Sun Jun 12 15:52:00 CDT 2016"),
        "APIEnabled" => false,
        "APIVersionsSupported" => array(),
        "LastAttempt" => 0,
        "LastUpdated" => 0,
        "Network" => "mainnet",
        "URL" => "https://dcr.ubiqsmart.com",
    ),
    "Kilo" => array(
        //"LaunchedEpoch" => strtotime("Tue Feb  7 17:00:00 CDT 2017"),
        "APIEnabled" => false,
        "APIVersionsSupported" => array(),
        "LastAttempt" => 0,
        "LastUpdated" => 0,
        "Network" => "testnet",
        "URL" => "https://teststakepool.decred.org",
    ),
    "Lima" => array(
        "APIEnabled" => false,
        "APIVersionsSupported" => array(),
        "LastAttempt" => 0,
        "LastUpdated" => 0,
        "Network" => "mainnet",
        "URL" => "https://ultrapool.eu",
    )
);

foreach ($spdata as $i => $d) {
    apcu_add("spcache-{$i}", $d);
}

switch ($_REQUEST["c"]) {
// clear cache
case "cc":
    if ($_SERVER["REMOTE_ADDR"] == "127.0.0.1" || $_SERVER["REMOTE_ADDR"] == "::1") {
        apcu_clear_cache();
        print "cache cleared\n";
    } else {
        print "unauthorized\n";
    }
    break;
case "dc":
    header("Content-Type: application/json");
    $count = downloadsCount();
    print json_encode(array("DownloadsCount", $count), JSON_NUMERIC_CHECK);
    break;
// downloads image cache
case "dic":
    header("Content-Type: image/svg+xml");
    $svg = downloadsImageCache();
    print $svg;
    break;
// get coin supply
case "gcs":
    header("Content-Type: application/json");
    $gscData = getCoinSupply();
    print json_encode($gscData, JSON_NUMERIC_CHECK|JSON_PRETTY_PRINT);
    break;
// get insight status
case "gis":
    header("Content-Type: application/json");
    $status = getInsightStatus();
    print $status;
    break;
// get stakepool data
case "gsd":
    header("Content-Type: application/json");
    getStakepoolData($spdata);
    $allpooldata = array();
    foreach (array_keys($spdata) as $i) {
        $pooldata = apcu_fetch("spcache-{$i}");
        // skip dcrstakes until they upgrade to the current API
        if ($pooldata["URL"] == "https://dcrstakes.com" && !in_array(STAKEPOOL_API_CURRENT_VERSION, $pooldata["APIVersionsSupported"])) {
            continue;
        }
        $allpooldata[$i] = $pooldata;
    }
    array_shuffle($allpooldata);
    print json_encode($allpooldata, JSON_NUMERIC_CHECK|JSON_PRETTY_PRINT);
    break;
default:
    print "unknown command";
}

function array_shuffle(&$array) {
    $keys = array_keys($array);

    shuffle($keys);

    foreach($keys as $key) {
        $new[$key] = $array[$key];
    }

    $array = $new;

    return true;
}

function debugLog($s) {
    if ($GLOBALS["debug"]) {
        error_log($s);
    }
}

function downloadsCount() {
    $cacheTTL = 4 * 60 * 60;

    // use the cached value if it is present
    $cachedDownloadsCount = apcu_fetch("downloadsCount");
    if (!empty($cachedDownloadsCount)) {
        return $cachedDownloadsCount;
    }

    // get a new value and cache it if it's good
    $newDownloadsCount = downloadsGetCount();
    if ($newDownloadsCount > 0) {
        $countString = round($newDownloadsCount / 1000) . "k";
        apcu_store("downloadsCount", $countString, $cacheTTL);
    }

    // return a default count from just before 0.8.0 if no cache
    // and couldn't get a new good count
    if (!$count) {
        return "61k";
    }
}

function downloadsFetchURL($URL) {
    $timeOut = 5;
    $c = curl_init($URL);
    curl_setopt($c, CURLOPT_USERAGENT, "decred/dcrweb bot");
    curl_setopt($c, CURLOPT_CONNECTTIMEOUT, $timeOut);
    curl_setopt($c, CURLOPT_RETURNTRANSFER, 1);
    curl_setopt($c, CURLOPT_TIMEOUT, $timeOut);
    $r = curl_exec($c);
    if ($r === false) {
        error_log("curl error: " . curl_error($c) . " (errno: " . curl_errno($c)  . ") while scraping {$URL}");
    }
    curl_close($c);

    return $r;
}

function downloadsGetCount() {
    $decred_binaries = downloadsFetchURL("https://api.github.com/repos/decred/decred-binaries/releases");
    $decred_release = downloadsFetchURL("https://api.github.com/repos/decred/decred-release/releases");

    if (!json_decode($decred_binaries, true)) {
        error_log("decred_binaries was not valid JSON: '{$decred_binaries}'");
        exit;
    }

    if (!json_decode($decred_release, true)) {
        error_log("decred_release was not valid JSON: '{$decred_release}'");
        exit;
    }

    $dataset = array(
        "db" => json_decode($decred_binaries, true),
        "dr" => json_decode($decred_release, true),
    );

    $count = 0;

    foreach (array_values($dataset) as $data) {
        foreach (array_values($data) as $release) {
            foreach (array_values($release) as $stats) {
                if (is_array($stats)) {
                    foreach ($stats as $id => $keys) {
                        if (is_array($keys)) {
                            $file = "";
                            foreach ($keys as $k => $v) {
                                if ($k == "name") {
                                    $file = $v;
                                }
                                if ($k == "download_count") {
                                    //print "$file => $v\n";
                                    $count = $count + $v;
                                }
                            }
                        }
                    }
                }
            }
        }
        debugLog("count $count");
    }

    return $count;
}

// XXX will need adjusting when integer count reaches 6 digits
function downloadsImageCache() {
    $cacheTTL = 4 * 60 * 60;
    $svgTemplate = '<svg xmlns="http://www.w3.org/2000/svg" width="128" height="20"><linearGradient id="b" x2="0" y2="100%"><stop offset="0" stop-color="#bbb" stop-opacity=".1"/><stop offset="1" stop-opacity=".1"/></linearGradient><mask id="a"><rect width="128" height="20" rx="3" fill="#fff"/></mask><g mask="url(#a)"><path fill="#555" d="M0 0h69v20H0z"/><path fill="#4c1" d="M69 0h59v20H69z"/><path fill="url(#b)" d="M0 0h128v20H0z"/></g><g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="11"><text x="34.5" y="15" fill="#010101" fill-opacity=".3">downloads</text><text x="34.5" y="14">downloads</text><text x="97.5" y="15" fill="#010101" fill-opacity=".3">__COUNT__ total</text><text x="97.5" y="14">__COUNT__ total</text></g></svg>';

    $curSVG = apcu_fetch("downloadsSVG");
    if (empty($curSVG)) {
        $count = downloadsGetCount();
        if (!$count) {
            error_log("downloadsGetCount returned '{$count}'");
            exit;
        }
        $countString = round($count / 1000) . "k";

        $newSVG = str_replace("__COUNT__", $countString, $svgTemplate);
        apcu_store("downloadsSVG", $newSVG, $cacheTTL);
        return $newSVG;
    } else {
        return $curSVG;
    }
}

function getCoinSupply() {
    $cacheTTL = 60;
    $timeOut = 3;
    $URL = "https://mainnet.decred.org/api/status?q=getCoinSupply";

    $curCoinSupply = apcu_fetch("gsc");
    if (empty($curCoinSupply)) {
        $c = curl_init($URL);
        curl_setopt($c, CURLOPT_USERAGENT, "decred/dcrweb bot");
        curl_setopt($c, CURLOPT_CONNECTTIMEOUT, $timeOut);
        curl_setopt($c, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($c, CURLOPT_TIMEOUT, $timeOut);
        $r = curl_exec($c);
        if ($r === false) {
            error_log("curl error: " . curl_error($c) . " (errno: " . curl_errno($c)  . ") while scraping {$URL}");
        } else {
            $jd = json_decode($r, true);
            if (!empty($jd)) {
                apcu_store("gsc", $jd["coinsupply"], $cacheTTL);
                $curCoinSupply = $jd["coinsupply"];
            } else {
                $curCoinSupply = "464594945383704";
            }
        }
        curl_close($c);
    }

    $airdrop = $premine = 840000;
    $coinSupplyAvailable = round($curCoinSupply / 100000000);
    $coinSupplyAfterBlock1 = $coinSupplyAvailable - $airdrop - $premine;
    $coinSupplyTotal = 21000000;

    $data = array(
        "PercentMined" => round($coinSupplyAvailable/$coinSupplyTotal*100, 1, PHP_ROUND_HALF_UP),
        "CoinSupplyMined" => $coinSupplyAvailable,
        "CoinSupplyMinedRaw" => $curCoinSupply,
        "CoinSupplyTotal" => $coinSupplyTotal,
        "Airdrop" => round($airdrop/$coinSupplyAvailable*100, 1, PHP_ROUND_HALF_UP),
        "Pos" => round($coinSupplyAfterBlock1*30/100/$coinSupplyAvailable*100, 1, PHP_ROUND_HALF_UP),
        "Pow" => round($coinSupplyAfterBlock1*60/100/$coinSupplyAvailable*100, 1, PHP_ROUND_HALF_UP),
        "Premine" => round($premine/$coinSupplyAvailable*100, 1, PHP_ROUND_HALF_UP),
        "Subsidy" => round($coinSupplyAfterBlock1*10/100/$coinSupplyAvailable*100, 1, PHP_ROUND_HALF_UP),
    );

    return $data;
}

function getInsightStatus() {
    $cacheTTL = 60;
    $timeOut = 3;
    $status = '{"info":{"blocks":"-"}}';
    $URL = "https://mainnet.decred.org/api/status";

    $curStatus = apcu_fetch("gsi");

    if (empty($curStatus)) {
        $c = curl_init($URL);
        curl_setopt($c, CURLOPT_USERAGENT, "decred/dcrweb bot");
        curl_setopt($c, CURLOPT_CONNECTTIMEOUT, $timeOut);
        curl_setopt($c, CURLOPT_RETURNTRANSFER, 1);
        curl_setopt($c, CURLOPT_TIMEOUT, $timeOut);
        $r = curl_exec($c);
        if ($r === false) {
            error_log("curl error: " . curl_error($c) . " (errno: " . curl_errno($c)  . ") while scraping {$URL}");
        } else {
            $jd = json_decode($r, true);
            if (!empty($jd)) {
                apcu_store("gsi", $r, $cacheTTL);
            }
            $status = $r;
        }
        curl_close($c);
    } else {
        $status = $curStatus;
    }

    return $status;
}

function getStakepoolStatsAPI($poolURL, $timeOut, $fields, $apiversion) {
    $empty = array();

    $apiURL = "$poolURL/api/v${apiversion}/stats";

    debugLog("GET $apiURL");
    $c = curl_init($apiURL);
    curl_setopt($c, CURLOPT_CONNECTTIMEOUT, $timeOut);
    curl_setopt($c, CURLOPT_RETURNTRANSFER, 1);
    curl_setopt($c, CURLOPT_SSL_VERIFYPEER, 0);
    // XXX if getstakeinfo isn't cached then this takes a long time
    // XXX should probably be parallelized
    curl_setopt($c, CURLOPT_TIMEOUT, $timeOut*3);
    $r = curl_exec($c);
    if ($r === false) {
        error_log("curl error: " . curl_error($c) . " (errno: " . curl_errno($c)  . ") while scraping $apiURL");
        // XXX should probably check to see if the curl error code shows a timeout occurred
        return array(1, $empty);
    }

    $httpCode = curl_getinfo($c, CURLINFO_HTTP_CODE);
    curl_close($c);

    // pools that haven't implemented this yet will likely 404 or possibly
    // redirect to an error page.
    if ($httpCode != 200) {
        debugLog("HTTP status code was $httpCode");
        return array(0, $empty);
    }

    $json = json_decode($r, true);
    if ($json === false) {
        debugLog("failed to parse $json");
        return array(0, $empty);
    } else {
        if ($json["status"] == "success") {
            foreach ($fields as $field) {
                if (!isset($json["data"][$field])) {
                    debugLog("missing field $field in response");
                    return array(0, $empty);
                }
            }
            return array(0, $json["data"]);
        } else if ($json["status"] == "error") {
            debugLog("API error: {$json["message"]}");
            return array(0, $empty);
        } else {
            debugLog("improper json response $r");
            return array(0 , $empty);
        }
    }
}

function getStakepoolData($spdata) {
    $fields = array(
            "Immature",
            "Live",
            "Voted",
            "Missed",
            "PoolFees",
            "ProportionLive",
            "UserCount",
            "UserCountActive",
    );

    $fieldtypes = array(
            "Immature" => "int",
            "Live" => "int",
            "Voted" => "int",
            "Missed" => "int",
            "PoolFees" => "float",
            "ProportionLive" => "float",
            "UserCount" => "int",
            "UserCountActive" => "int",
    );

    $interval = 20 * 60;
    $timeOut = 2;

    foreach (array_keys($spdata) as $i) {
        $cachedData = apcu_fetch("spcache-{$i}");
        $stats = array();

        if (isset($cachedData["LastUpdated"])
            && time() - $cachedData["LastUpdated"] > $interval
            && time() - $cachedData["LastAttempt"] > $interval) {

            // if a pool's URL changed then we need to force usage of it
            if ($cachedData["URL"] != $spdata[$i]["URL"]) {
                $cachedData["URL"] = $spdata[$i]["URL"];
            }

            debugLog("updating {$cachedData["URL"]}");
            $cachedData["LastAttempt"] = time();

            // need to force these keys into existence for upgrades but we don't
            // want to overwrite a cached true value
            if (!isset($cachedData["APIEnabled"])) {
                $cachedData["APIEnabled"] = false;
            }

            if (!isset($cachedData["APIVersionsSupported"])) {
                $cachedData["APIVersionsSupported"] = 0;
            }

            if (!isset($cachedData["Network"])) {
                $cachedData["Network"] = $spdata[$i]["Network"];
            }

            // first try current API version
            list ($timedOut, $stats) = getStakepoolStatsAPI($cachedData["URL"], $timeOut, $fields, STAKEPOOL_API_CURRENT_VERSION);

            // if current API version worked then note that
            if (!empty($stats)) {
                debugLog("got stats via APIv" . STAKEPOOL_API_CURRENT_VERSION . " from {$cachedData["URL"]}");
                $cachedData["APIEnabled"] = true;
                $cachedData["APIVersionsSupported"] = array(
                    STAKEPOOL_API_INITIAL_VERSION,
                    STAKEPOOL_API_CURRENT_VERSION
                );
            }

            // fall back to old API version but only if we didn't timeout
            if (empty($stats)) {
                if (!$timedOut) {
                    list ($timedOut, $stats) = getStakepoolStatsAPI($cachedData["URL"], $timeOut, $fields, 1);
                    // if API v1 worked then note that
                    if (!$timedOut && !empty($stats)) {
                        debugLog("got stats via APIv " . STAKEPOOL_API_INITIAL_VERSION . " from {$cachedData["URL"]}");
                        $cachedData["APIEnabled"] = true;
                        $cachedData["APIVersionsSupported"] = array(
                            STAKEPOOL_API_INITIAL_VERSION
                        );
                    }
                }
            }

            // update the LastAttempt and continue on failure
            if (empty($stats)) {
                debugLog("no stats available from {$cachedData["URL"]}");
                apcu_store("spcache-{$i}", $cachedData);
                continue;
            }

            foreach ($stats as $k => $v) {
                if ($v === "") {
                    $stats[$k] = -1;
                }
            }

            $cachedData["LastUpdated"] = time();
            foreach ($fields as $field) {
                // don't cache failed reads
                if (array_key_exists($field, $cachedData) && $cachedData[$field] != -1 && $stats[$field] == -1) {
                    debugLog("skipping caching empty value for $field");
                    continue;
                }

                if (isset($stats[$field])) {
                    $cachedData[$field] = $stats[$field];
                }
            }

            apcu_store("spcache-{$i}", $cachedData);

            debugLog("successfully updated {$cachedData["URL"]}");
        } else {
            debugLog("no updates required for {$cachedData["URL"]}");
        }
    }
}
?>
